src/lib.rs

```diff
@@ -875,11 +901,35 @@ impl<T> core::future::Future for Receiver<T> {
             }
             // The sender sent the message.
             MESSAGE => {
-                channel.state.store(DISCONNECTED, SeqCst);
+                // ORDERING: the sender has been `mem::forget`-ed so this update only needs to be
+                // visible to us
+                channel.state.store(DISCONNECTED, Relaxed);
                 Poll::Ready(Ok(unsafe { channel.take_message() }))
             }
             // The sender was dropped before sending anything, or we already received the message.
             DISCONNECTED => Poll::Ready(Err(RecvError)),
+            // The sender has written the message and is currently reading the waker from a
+            // previous poll. This means the message is already written. But we need to loop here
+            // until we observe the MESSAGE state, so we can flip it to DISCONNECTED, to avoid
+            // reading the message twice.
+            // We busy loop here since we know the sender is done very soon. Waking ourselves up
+            // and return Poll::Pending here would also result in a busy loop that is likely
+            // more expensive.
+            UNPARKING => loop {
```

Comment:
This entire branch was missing here. The poll method is the only receiving method that can observe `UNPARKING` in the outermost load. Since a previous poll can set the `RECEIVING` state and the sender can then have progressed into `UNPARKING`.

Reply:
This loop looks mostly correct, but I would just load the state in a loop and take the message if we wind up in the `MESSAGE` state, and return an error if we end up in the `DISCONNECTED` state. There's no need to make this state change occur in a single atomic step so the store can happen later.

Reply:
Fixed